### PR TITLE
Cr refine push

### PR DESCRIPTION
**Pull request by fyin1**

### CODE REVIEW

File: hypervisor/arch/x86/vmx.c [hunk 1]
```diff
@@ -374,22 +408,20 @@ int vmx_write_cr0(struct vcpu *vcpu, uint64_t cr0)
 	uint32_t entry_ctrls;
 	bool paging_enabled = !!(context->cr0 & CR0_PG);
 
-	if ((cr0 & (cr0_always_off_mask | CR0_RESERVED_MASK)) != 0UL) {
-		pr_err("Not allow to set always off / reserved bits for CR0");
+	if (is_cr0_write_valid(vcpu, cr0) == false) {
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vmx.c]: is_cr0_write_valid returns Bool. So you don't need " ... == false". I think we can directly use if ( !xxxx). Please check with Junjie

File: hypervisor/arch/x86/vmx.c [hunk 2]
```diff
@@ -374,22 +408,20 @@ int vmx_write_cr0(struct vcpu *vcpu, uint64_t cr0)
...
+		pr_err("Invalid cr0 write operation from guest");
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vmx.c]: remove this pr_eff... This is normal...  use pr_dbg ?

File: hypervisor/include/arch/x86/cpu.h [hunk 3]
```diff
@@ -83,13 +83,16 @@
...
 #define CR4_VMXE                (1UL<<13)	/* VMX enable */
 #define CR4_SMXE                (1UL<<14)	/* SMX enable */
+#define CR4_FSGSBASE            (1UL<<16)	/* RD(FS|GS|FS)BASE inst */
 #define CR4_PCIDE               (1UL<<17)	/* PCID enable */
 #define CR4_OSXSAVE             (1UL<<18)
+/* XSAVE and Processor Extended States enable bit */
 #define CR4_SMEP                (1UL<<20)
 #define CR4_SMAP                (1UL<<21)
-/* XSAVE and Processor Extended States enable bit */
+#define CR4_PKE                 (1UL<<22)	/* Protect-key-enable */
```
Comment (dongyaozu) [inline hypervisor/include/arch/x86/cpu.h]: << 22U. Change the CR4 bit definition too.

File: hypervisor/arch/x86/vmx.c [hunk 4]
```diff
@@ -543,19 +563,8 @@ int vmx_write_cr4(struct vcpu *vcpu, uint64_t cr4)
...
-	if((cr4 & cr4_always_off_mask) != 0U) {
-		pr_err("Not allow to set reserved/always off bits for CR4");
-		vcpu_inject_gp(vcpu, 0U);
-		return 0;
-	}
-
-	/* Do NOT support nested guest */
-	if ((cr4 & CR4_VMXE) != 0UL) {
-		pr_err("Nested guest not supported");
+	if (is_cr4_write_valid(vcpu, cr4) == false) {
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vmx.c]: no need of == false.

File: hypervisor/arch/x86/vmexit.c [hunk 5]
```diff
@@ -281,6 +281,15 @@ int cr_access_vmexit_handler(struct vcpu *vcpu)
 		break;
 	case 0x08U:
 		/* mov to cr8 */
+		/* According to SDM 6.15 "Exception and interrupt Reference":
+		 *
+		 * set reserved bit in CR8 causes GP to guest
+		 */
+		if (*regptr & ~0xFUL) {
+			pr_err("Invalid cr8 write operation from guest");
```
Comment (dongyaozu) [inline hypervisor/arch/x86/vmexit.c]: no pr_err, dbg is fine.

File: hypervisor/arch/x86/vmx.c (hunk 1 above)
Comment (fyin1) [reply hypervisor/arch/x86/vmx.c]: OK. Will check with Junjie and update the patch.

File: hypervisor/arch/x86/vmx.c (hunk 2 above)
Comment (fyin1) [reply hypervisor/arch/x86/vmx.c]: OK.

File: hypervisor/include/arch/x86/cpu.h (hunk 3 above)
Comment (fyin1) [reply hypervisor/include/arch/x86/cpu.h]: Yes. I noticed this one and planned to have separate patch for this change.
I will handle this in new push of this patch series.

File: hypervisor/arch/x86/vmx.c (hunk 4 above)
Comment (fyin1) [reply hypervisor/arch/x86/vmx.c]: OK.

File: hypervisor/arch/x86/vmexit.c (hunk 5 above)
Comment (fyin1) [reply hypervisor/arch/x86/vmexit.c]: OK.

**Comment by lijinxia** [comment]
Verified+1

**Pull request merged**